FLEXIBLE_INDEXES_NOTES.md

```diff
@@ -0,0 +1,124 @@
+# Proposal: Xarray flexible indexes refactoring
+
+Current status: https://github.com/pydata/xarray/projects/1
+
+TODO: check if the project is up-to-date and update it if necessary.
+
+## 1. Data Model
+
+Indexes are used in Xarray to extract data from Xarray objects using coordinate labels instead of using integer array indices. Although the indexes used in an Xarray object can be accessed (or built on-the-fly) via public methods like `to_index()` or properties like `indexes`, those are mainly used internally.
+
+The goal of this project is to make those indexes 1st-class citizens of Xarray's data model. As such, indexes should clearly be separated from Xarray coordinates with the following relationships:
+
+- Index -> Coordinate: one-to-many
+- Coordinate -> Index: one-to-zero-or-one
+
+An index may be built from one or more coordinates. However, each coordinate must relate to one index at most. Additionally, a coordinate may not be tied to any index.
```

Comment:
I think this is a reasonable idea, but perhaps could be implemented via a "meta" index that wraps multiple other indexes internally.

`method='nearest'` in `pd.Index` is already sort of like this: it uses `np.searchsorted()` rather than the hashtable used for other look-ups.

The advantage of only associating one index with a coordinate is that looking up which index to use from an indexing operation is unambiguous and fast (using a dict).

Reply:
Yes I agree that is a relevant example. Another use case could be some tree-based indexes that only support point-wise indexing while others would allow providing slices to `.sel()` (range-based queries of irregular data).

I like the "meta" index idea. It would prevent having to handle complicated situations in Xarray such as checking for potential ambiguous situations where two indexes assigned to the same coordinate both support the same operations. It would also keep the API simple and less confusing for users.